File: source/Tilemap.lua
-- SPDX-FileCopyrightText: 2022-present Didier Malenfant <coding@malenfant.net>
--
-- SPDX-License-Identifier: MIT

import "CoreLibs/object"

dm = dm or {}

class('LuaTilemap', { }, dm).extends()

local LuaTilemap <const> = dm.LuaTilemap
local gfx <const> = playdate.graphics
local display_width <const>, display_height <const> = playdate.display.getSize()

function LuaTilemap:init()
    LuaTilemap.super.init(self)

    self.image_table = nil
    self.map = nil

    self.height = 0
    self.width = 0

    self.tile_width = 0
    self.tile_height = 0
end

-- Sets the tilemap’s playdate.graphics.imagetable to table, a playdate.graphics.imagetable.
function LuaTilemap:setImageTable(table)
    self.image_table = table

    local image <const> = self.image_table:getImage(1)
    self.tile_width = image.width
    self.tile_height = image.height
end

-- Sets the tilemap’s width to width, then populates the tilemap with data, which should be
-- a flat, one-dimensional array-like table containing index values to the tilemap’s imagetable.
function LuaTilemap:setTiles(data, width)
    self.map = data
    self.width = width
end

-- Returns data, width
-- data is a flat, one-dimensional array-like table containing index values to the tilemap’s imagetable.
-- width is the width of the tile map, in number of tiles.
function LuaTilemap:getTiles()
    return self.data, self.width
end

-- Draws the tile map at screen coordinate (x, y).
-- sourceRect, if specified, will cause only the part of the tilemap within sourceRect to be drawn.
function LuaTilemap:draw(x, y, _sourceRect)
    local image_width, image_height <const> = self:getTileSize()

    local draw_x = x
    local draw_y = y

    local tile_x = 0
    local tile_y = 0

    if draw_x < -16 then
        draw_x = -(-draw_x % image_width)
        tile_x = math.floor((draw_x - x) / image_width)
    end

    if tile_x >= self.width then
        return
    end

    if draw_y < -16 then
        draw_y = -(-draw_y % image_height)
        tile_y = math.floor((draw_y - y) / image_height)
    end

    if tile_y >= self.height then
        return
    end

    local current_tile_x = tile_x
    local current_tile_y = tile_y

    local current_draw_x = draw_x
    local current_draw_y = draw_y

    local width <const> = self.width
    local height <const> = self.height
    local map <const> = self.map
    local table <const> = self.image_table

    local tile_index = (current_tile_y * width) + current_tile_x
    while((current_draw_y < display_height) and (current_tile_y < height)) do
        local next_tile_index_offset = width

        while((current_draw_x < display_width) and (current_tile_x < width)) do
            local image_index = map[tile_index]
            if image_index ~= 0 then
                local image = table:getImage(image_index)
                if image ~= nil then
                    image:draw(current_draw_x, current_draw_y)
                end
            end

            next_tile_index_offset = next_tile_index_offset - 1
            tile_index = tile_index + 1

            current_draw_x = current_draw_x + image_width
            current_tile_x = current_tile_x + 1
        end

        tile_index = tile_index + next_tile_index_offset

        current_tile_x = tile_x
        current_tile_y = current_tile_y + 1

        current_draw_x = draw_x
        current_draw_y = current_draw_y + image_height
    end

    local debug = false
    if debug then
        local text = "xy: "..tostring(x)..' '..tostring(y)
        local text_width, text_height = gfx.getTextSize(text)
        gfx.fillRect(0, 30, text_width, text_height)
        gfx.setImageDrawMode("fillWhite")
        gfx.drawText(text, 0, 30)

        text = "draw: "..tostring(draw_x)..' '..tostring(draw_y)
        text_width, text_height = gfx.getTextSize(text)
        gfx.fillRect(0, 50, text_width, text_height)
        gfx.drawText(text, 0, 50)

        text = "tile: "..tostring(tile_x)..' '..tostring(tile_y)
        text_width, text_height = gfx.getTextSize(text)
        gfx.fillRect(0, 70, text_width, text_height)
        gfx.drawText(text, 0, 70)
    end
end

-- Sets the index of the tile at tilemap position (x, y). index is the (1-based) index of the image
-- in the tilemap’s playdate.graphics.imagetable.
function LuaTilemap:setTileAtPosition(x, y, index)
    self.map[(x - 1) + ((y - 1) * self.width)] = index
end

-- Returns the image index of the tile at the given x and y coordinate. If x or y is out of bounds, returns nil.
function LuaTilemap:getTileAtPosition(x, y)
    return self.map[x + ((y - 1) * self.width)]
end

-- Sets the tilemap’s width and height, in number of tiles.
function LuaTilemap:setSize(width, height)
    self.width = width
    self.height = height
    self.map = {}
end

-- Returns the size of the tile map, in tiles, as a pair, (width, height).
function LuaTilemap:getSize()
    return self.width, self.height
end

-- Returns the size of the tilemap in pixels; that is, the size of the image multiplied by the number of
-- rows and columns in the map. Returns multiple alues (width, height).
function LuaTilemap:getPixelSize()
    local image_width, image_height <const> = self:getTileSize()
    return self.width * image_width, self.height * image_height
end

-- Returns multiple values (width, height), the pixel width and height of an individual tile.
function LuaTilemap:getTileSize()
    return self.tile_width, self.tile_height
end

File: Tilemap/Tilemap.c
// SPDX-FileCopyrightText: 2022-present Didier Malenfant <coding@malenfant.net>
//
// SPDX-License-Identifier: MIT

#include "Tilemap/Tilemap.h"

#define DM_LOG_ENABLE
#include "pdbase/pdbase.h"

// -- Forward declaration
static const lua_reg tilemapClass[];

// -- Constants
#define CLASSNAME_TILEMAP "dm.Tilemap"

// -- Tilemap class
typedef struct {
    LCDBitmapTable* image_table;

    int height;
    int width;

    int tile_width;
    int tile_height;

    int nb_of_tiles;
    LCDBitmap** tiles;
    
    uint16_t* map;
} Tilemap;

// -- Get an argument as a Tilemap class
#define GET_TILEMAP_ARG(index)    pd->lua->getArgObject(index, CLASSNAME_TILEMAP, NULL);

// -- Register the class
extern void register_Tilemap(PlaydateAPI* api)
{
    const char* err = NULL;
    
    // -- Register Tilemap
    if (!pd->lua->registerClass(CLASSNAME_TILEMAP, tilemapClass, NULL, 0, &err))
    {
        DM_LOG("dm.Tilemap: Failed to register the Tilemap class (%s).", err);
        return;
    }
}

// -- Allocate a new tilemap
int tilemapNew(lua_State* L)
{
    Tilemap* this = dmMemoryCalloc(1, sizeof(Tilemap));
    if (this == NULL) {
        return 0;
    }

    const char* path = pd->lua->getArgString(1);
    if (path == NULL) {
        DM_LOG("Tilemap: Error getting tilemap path argument.");
        return 0;
    }

    const char* err = NULL;
    this->image_table = pd->graphics->loadBitmapTable(path, &err);
    if (this->image_table == NULL) {
        DM_LOG("Tilemap: Error loading image table '%s' (%s).", path, (err != NULL) ? err : "Unknown Error");
        return 0;
    }

    this->nb_of_tiles = 0;
    for (; ;++this->nb_of_tiles) {
        LCDBitmap* bitmap = pd->graphics->getTableBitmap(this->image_table, this->nb_of_tiles);
        if (bitmap == NULL) {
            break;
        }
    }

    DM_LOG("Found %d tiles.", this->nb_of_tiles);
    this->tiles = dmMemoryCalloc(this->nb_of_tiles, sizeof(LCDBitmap*));

    for (int index = 0; index < this->nb_of_tiles;++index) {
        this->tiles[index] = pd->graphics->getTableBitmap(this->image_table, index);
    }
    
    this->height = 0;
    this->width = 0;

    LCDBitmap* bitmap = pd->graphics->getTableBitmap(this->image_table, 1);
    if (bitmap == NULL) {
        DM_LOG("Tilemap: Error getting bitmap from image table '%s'.", path);
        return 0;
    }

    int width, height;
    pd->graphics->getBitmapData(bitmap, &width, &height, NULL, NULL, NULL);

    this->tile_height = width;
    this->tile_width = height;
    
    pd->graphics->freeBitmap(bitmap);

    this->map = NULL;

    pd->lua->pushObject(this, CLASSNAME_TILEMAP, 0);

    return 1;
}
 
// -- Delete the tilemap
int tilemapDelete(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }
    
    if (this->image_table != NULL) {
        pd->graphics->freeBitmapTable(this->image_table);
        this->image_table = NULL;
    }
    
    if (this->tiles != NULL) {
        dmMemoryFree(this->tiles);
        this->image_table = NULL;
        this->nb_of_tiles = 0;
    }
    
    if (this->map != NULL) {
        dmMemoryFree(this->map);
        this->map = NULL;
    }
    
    dmMemoryFree(this);
    
    return 0;
}

// -- Draws the tile map at screen coordinate (x, y).
// -- sourceRect, if specified, will cause only the part of the tilemap within sourceRect to be drawn.
// function Tilemap:draw(x, y, _sourceRect)
int tilemapDraw(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }

    if (this->map == NULL) {
        DM_LOG("Tilemap: Size of tilemap not set before draw().");
        return 0;
    }
    
    int image_width = this->tile_width;
    int image_height = this->tile_height;

    int x = pd->lua->getArgInt(2);
    int y = pd->lua->getArgInt(3);

    int draw_x = x;
    int draw_y = y;

    int tile_x = 0;
    int tile_y = 0;

    if (draw_x < -image_width) {
        draw_x = -(-draw_x % image_width);
        tile_x = (int)((draw_x - x) / image_width);
    }

    if (tile_x >= this->width) {
        return 0;
    }

    if (draw_y < -image_height) {
        draw_y = -(-draw_y % image_height);
        tile_y = (int)((draw_y - y) / image_height);
    }

    if (tile_y >= this->height) {
        return 0;
    }

    int current_tile_x = tile_x;
    int current_tile_y = tile_y;

    int current_draw_x = draw_x;
    int current_draw_y = draw_y;

    int width = this->width;
    int height = this->height;
    uint16_t* tilemap = this->map;
    LCDBitmapTable* table = this->image_table;

    int display_width = pd->display->getWidth();
    int display_height = pd->display->getHeight();

    int tilemap_index = (current_tile_y * width) + current_tile_x;
    while((current_draw_y < display_height) && (current_tile_y < height)) {
        int next_tilemap_index_offset = width;

        while((current_draw_x < display_width) && (current_tile_x < width)) {
            int tile_index = tilemap[tilemap_index] - 1;
            if (tile_index != 0) {
                /*
                if (tile_index >= this->nb_of_tiles) {
                    DM_LOG("Tilemap: Invalid tile index %d.", tile_index);
                    return 0;
                }*/
                
                LCDBitmap* bitmap = this->tiles[tile_index];
                //LCDBitmap* bitmap = pd->graphics->getTableBitmap(this->image_table, tile_index);
                //if (bitmap != NULL) {
                    pd->graphics->drawBitmap(bitmap, current_draw_x, current_draw_y, kBitmapUnflipped);
                    //pd->graphics->drawBitmap(otherBitmap, current_draw_x, current_draw_y, kBitmapUnflipped);
                //}
            }

            next_tilemap_index_offset -= 1;
            tilemap_index += 1;

            current_draw_x += image_width;
            current_tile_x += 1;
        }

        tilemap_index += next_tilemap_index_offset;

        current_tile_x = tile_x;
        current_tile_y += 1;

        current_draw_x = draw_x;
        current_draw_y += image_height;
    }

    return 0;
}

// -- Sets the index of the tile at tilemap position (x, y). index is the (1-based) index of the image
// -- in the tilemap’s playdate.graphics.imagetable.
// function Tilemap:setTileAtPosition(x, y, index)
int tilemapSetTileAtPosition(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }
    
    if (this->map == NULL) {
        DM_LOG("Tilemap: Size of tilemap not set before setTileAtPosition().");
        return 0;
    }
    
    int x = pd->lua->getArgInt(2);
    int y = pd->lua->getArgInt(3);
    
    if ((x < 1) || (x > this->width) || (y < 1) || (y > this->height)) {
        DM_LOG("Tilemap: Out of bounds values %d,%s for getTileAtPosition.", x, y);
        return 0;
    }

    int tilemap_index = pd->lua->getArgInt(4);
    if (tilemap_index < 0) {
        DM_LOG("Tilemap: Out of bounds tile index %d for getTileAtPosition.", tilemap_index);
        return 0;
    }
    
    this->map[((y - 1) * this->width) + (x - 1)] = tilemap_index;

    return 0;
}

// -- Returns the image index of the tile at the given x and y coordinate. If x or y is out of bounds, returns nil.
// function Tilemap:getTileAtPosition(x, y)
int tilemapGetTileAtPosition(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }

    if (this->map == NULL) {
        DM_LOG("Tilemap: Size of tilemap not set before getTileAtPosition().");
        return 0;
    }

    int x = pd->lua->getArgInt(2);
    int y = pd->lua->getArgInt(3);

    if ((x < 1) || (x > this->width) || (y < 1) || (y > this->height)) {
        DM_LOG("Tilemap: Out of bounds values %d,%s for getTileAtPosition.", x, y);
        return 0;
    }

    pd->lua->pushInt(this->map[((y - 1) * this->width) + (x - 1)]);

    return 1;
}

// -- Sets the tilemap’s width and height, in number of tiles.
// function Tilemap:setSize(width, height)
int tilemapSetSize(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }

    if (this->map != NULL) {
        dmMemoryFree(this->map);
        this->map = NULL;
    }
    
    this->width = pd->lua->getArgInt(2);
    this->height = pd->lua->getArgInt(3);

    if ((this->width == 0) || (this->width > 2048) || (this->height == 0) || (this->height > 2048)) {
        DM_LOG("Tilemap: Trying to set an invalid size of %dx%d.", this->width, this->height);
        return 0;
    }

    this->map = dmMemoryCalloc(this->width * this->height, sizeof(uint16_t));

    return 0;
}

// -- Returns the size of the tile map, in tiles, as a pair, (width, height).
// function Tilemap:getSize()
int tilemapGetSize(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }
    
    pd->lua->pushInt(this->width);
    pd->lua->pushInt(this->height);
    
    return 2;
}

// -- Returns the size of the tilemap in pixels; that is, the size of the image multiplied by the number of
// -- rows and columns in the map. Returns multiple values (width, height).
// function Tilemap:getPixelSize()
int tilemapGetPixelSize(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }

    pd->lua->pushInt(this->width * this->tile_width);
    pd->lua->pushInt(this->height * this->tile_height);

    return 2;
}

// -- Returns multiple values (width, height), the pixel width and height of an individual tile.
// function Tilemap:getTileSize()
int tilemapGetTileSize(lua_State* L)
{
    Tilemap* this = GET_TILEMAP_ARG(1);
    if(this == NULL) {
        DM_LOG("Tilemap: Error getting 'self' argument.");
        return 0;
    }
    
    pd->lua->pushInt(this->tile_width);
    pd->lua->pushInt(this->tile_height);
    
    return 2;
}

static const lua_reg tilemapClass[] = {
    { "new", tilemapNew },
    { "__gc", tilemapDelete },
    
    { "draw", tilemapDraw },
    { "setTileAtPosition", tilemapSetTileAtPosition },
    { "getTileAtPosition", tilemapGetTileAtPosition },
    { "setSize", tilemapSetSize },
    { "getSize", tilemapGetSize },
    { "getPixelSize", tilemapGetPixelSize },
    { "getTileSize", tilemapGetTileSize },
    
    { NULL, NULL }
};
